### PR TITLE
Parent Orchestration Access For SubOrchestration

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary

## What changed?
- Added `ParentOrchestrationInstance` type to expose parent orchestration information in sub-orchestrations
- Added abstract `parent` property to `OrchestrationContext` base class
- Implemented `parent` getter in `RuntimeOrchestrationContext` that extracts parent info from `ExecutionStartedEvent`
- Updated `orchestration-executor.ts` to extract `ParentInstanceInfo` from protobuf events
- Extended `newExecutionStartedEvent` helper to support `parentInstance` parameter for testing
- Exported `ParentOrchestrationInstance` type from package index

## Why is this change needed?
- Feature parity with durabletask-dotnet SDK which exposes `context.Parent` for sub-orchestrations
- Enables orchestrations to know if they are running as a sub-orchestration and access parent's name/instanceId
- Useful for logging, debugging, and building orchestration hierarchies

## Issues / work items
- Resolves #[TBD - add issue number]
- Related to feature parity initiative between JS and .NET SDKs

---

# Project checklist
- [x] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `CHANGELOG.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [ ] Breaking change?
  - [ ] If yes:
    - Impact: None - this is an additive change adding a new optional property
    - Migration guidance: N/A

---

# AI-assisted code disclosure (required)

## Was an AI tool used? (select one)
- [ ] No
- [x] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [ ] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot (Claude Opus 4.5)
- AI-assisted areas/files:
  - `src/types/parent-orchestration-instance.type.ts` (new file)
  - `src/task/context/orchestration-context.ts` (added abstract property)
  - `src/worker/runtime-orchestration-context.ts` (implemented parent getter)
  - `src/worker/orchestration-executor.ts` (extract parent from events)
  - `src/utils/pb-helper.util.ts` (test helper update)
  - `src/index.ts` (export)
  - `test/parent-orchestration-instance.spec.ts` (new unit tests)
  - `test/e2e-azuremanaged/orchestration.spec.ts` (new E2E tests)
  - `docs/FEATURE_PARITY.md` (documentation update)
- What you changed after AI output: Reviewed all code, verified proto API usage, validated test assertions

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing

## Automated tests
- Result: Passed

### Unit tests (5 new tests)
```
PASS test/parent-orchestration-instance.spec.ts
  ✓ should return undefined for parent when orchestration is not a sub-orchestration
  ✓ should return parent instance info when orchestration is a sub-orchestration
  ✓ should preserve parent info during replay
  ✓ should make parent info available in generator orchestrations
  ✓ should return parent info that can be used for debugging/logging

Test Suites: 10 passed, 10 total
Tests:       134 passed, 134 total
```

### E2E tests (2 new tests)
```
Test Suites: 4 passed, 4 total
Tests:       4 skipped, 49 passed, 53 total
```

## Manual validation (only if runtime/behavior changed)
- Environment: Windows 11, Node.js 22.x, DTS Emulator (Docker)
- Steps + observed results:
  1. Started DTS emulator via Docker
  2. Ran E2E test suite with `npm run test:e2e:azuremanaged:internal`
  3. Parent orchestration tests passed - child orchestration correctly received parent name, instanceId, and taskScheduledId
- Evidence: Test logs show parent info correctly extracted from `ExecutionStartedEvent`

---

# Notes for reviewers
- The `taskScheduledId` field is included in the `ParentOrchestrationInstance` interface (optional). This differs slightly from .NET which doesn't expose it publicly, but the proto supports it and it can be useful for debugging.
- The parent info is set by the backend (DTS/sidecar), not by the SDK when scheduling sub-orchestrations - this matches .NET behavior.
- Reference implementation: `durabletask-dotnet/src/Worker/Core/Shims/TaskOrchestrationContextWrapper.cs`

### CODE REVIEW

File: packages/durabletask-js/test/parent-orchestration-instance.spec.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      newExecutionStartedEvent(name, TEST_INSTANCE_ID, undefined, {
+        name: PARENT_ORCHESTRATOR_NAME,
+        instanceId: PARENT_INSTANCE_ID,
+        taskScheduledId: PARENT_TASK_SCHEDULED_ID,
+      }),
+    ];
+
+    // Import timer events
+    const { newTimerCreatedEvent, newTimerFiredEvent } = require("../src/utils/pb-helper.util");
+
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply packages/durabletask-js/test/parent-orchestration-instance.spec.ts]: done

File: packages/durabletask-js/test/parent-orchestration-instance.spec.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        taskScheduledId: PARENT_TASK_SCHEDULED_ID,
+      }),
+    ];
+
+    const executor = new OrchestrationExecutor(registry, testLogger);
+    const result = await executor.execute(TEST_INSTANCE_ID, [], newEvents);
+
+    // Verify parent instance info is accessible in generator functions
+    expect(capturedParent).toBeDefined();
+    expect(capturedParent!.name).toEqual(PARENT_ORCHESTRATOR_NAME);
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply packages/durabletask-js/test/parent-orchestration-instance.spec.ts]: ???

**Review by kaibocai — APPROVED** [review]
